marker_api/src/diagnostic.rs

```diff
@@ -13,24 +15,78 @@ use crate::{
 /// the actual display depends on the driver.
 pub struct DiagnosticBuilder<'ast> {
     lint: &'static Lint,
+    node: EmissionNodeId,
     msg: String,
-    node: EmissionNode,
-    span: Span<'ast>,
+    span: Option<Span<'ast>>,
     parts: Vec<DiagnosticPart<String, Span<'ast>>>,
+    /// This flag indicates, if the diagnostic messages should actually be emitted
+    /// at the end. This might be false, if the lint is allowed at the emission location.
+    emit_lint: bool,
 }
 
 #[allow(clippy::needless_pass_by_value)] // `&impl ToString` doesn't work
 impl<'ast> DiagnosticBuilder<'ast> {
-    pub(crate) fn new(lint: &'static Lint, node: EmissionNode, msg: String, span: Span<'ast>) -> Self {
+    pub(crate) fn new_dummy(lint: &'static Lint, node: EmissionNodeId) -> Self {
+        Self {
+            lint,
+            node,
+            msg: String::new(),
+            span: None,
+            parts: vec![],
+            emit_lint: false,
+        }
+    }
+
+    pub(crate) fn new(lint: &'static Lint, node: EmissionNodeId, msg: String, span: Span<'ast>) -> Self {
         Self {
             lint,
             msg,
             node,
-            span,
+            span: Some(span),
             parts: vec![],
+            emit_lint: true,
         }
     }
 
+    /// This function sets the main message of the created diagnostic.
+    ///
+    /// From rustc a lint emission would look like this:
+    /// ```text
+    ///  warning: <lint message>                <-- The message that is set by this function
+    ///  --> path/file.rs:1:1
+    ///   |
+    /// 1 | expression
+    ///   | ^^^^^^^^^^
+    ///   |
+    /// ```
+    pub fn set_main_message(&mut self, msg: impl ToString) -> &mut Self {
+        if self.emit_lint {
+            self.msg = msg.to_string();
+        }
+        self
+    }
+
+    /// This function sets the main [`Span`] of the created diagnostic.
+    /// [`AstContext::emit_lint`] will by default use the [`Span`] of the given
+    /// [`EmissionNode`].
+    ///
+    /// From rustc a lint emission would look like this:
+    /// ```text
+    ///  warning: <lint message>
+    ///  --> path/file.rs:1:1
+    ///   |
+    /// 1 | node
+    ///   | ^^^^                 <-- The main span set by this function
+    ///   |
+    /// ```
+    pub fn set_main_span(&mut self, span: &Span<'ast>) -> &mut Self {
```

Comment:
I wonder if this could have the same `impl EmissionNode` type that the constructor `emit_lint` function.

For example, [`darling::Error::with_span`](https://docs.rs/darling/latest/darling/struct.Error.html#method.with_span) uses a trait called `Spanned`, which is implemented for every `syn` AST node that has a `Span` and for `Span` itself. Could we use a similar approach? This would allow passing either the node, or the span

Reply:
That's a good idea :+1:

I have a question in this regard. Should the `Spanned::span` function replace the `span()` function of nodes, which already have one, or should it be an addition? Meaning:

```rs
struct Node {...}

impl Node {
    // Should this one be removed?
    // vvv
    fn span(&self) -> Span {...}
}

impl Spanned for Node {
	fn span(&self) -> Span {...}
}
```

Not having the function means that a user always has to import `Spanned` or `marker_api::prelude::*`. While having it, means that we have some code duplication. (Even if it's macro generated).

I'm always not 100% sure what's the best approach. I'm guessing, only providing the trait implementation, but some more input would be appreciated :)

Reply:
+
I admit, that it's not clear to me either. I did it once on my experience where I created a trait, but also added the inherent impl to the type that just forwards the call to the trait impl. But in that case I only had one trait.

But here the question is whether it makes sense to have this in general for all traits that are already present in `marker` as well.

There are several degrees of following/not-following this approach.
1. We can make all traits private, and make all the traits methods inherent impl methods of the types. This significantly reduces the API surface of marker, meaning more freedom to `marker_api`'s code, but introduces boilerplate (even macro-generated one), and users lack generic traits for the purposes of their own utils.
2. We make all traits public, and don't provide the inherent impls. This hurts the user experience, because users need to import the traits. If there is a `prelude` module that imports all traits, it reduces the problem. In this case we expect users to always have a `use marker_api::prelude::*` at the top.
3. We make all traits public and provide the inherent impls. Users never need to use the `prelude`, but we are left with writing boilerplate inherent impls.
4. We add inherent impls only for some widely-used trait methods, meaning users *mostly* don't need to use `prelude`, but for some advanced stuff and generic programming - definitely.

I don't know what's best, because there isn't an obviously correct choice. Subjectively I'd lean to 2 and require users to always have a `prelude` import. 

<sub>I also found [this cool proc macro](https://github.com/dtolnay/inherent) from David Tolnay.</sub>

Reply:
I think we should definitely provide public traits. So the decision is between 2, 3 and 4. I'm leaning towards 4. for really common traits, like `Spanned` would be.

---

I threw our conversation at ChatGBT and the response wasn't clear either. It basically listed the mentioned tradeoffs. It made one clear statement (It's ChatGBT, so take this with a grain of salt):
> Given your leaning towards Approach 2 (requiring trait imports), this approach seems reasonable, especially if it aligns with established practices in the Rust ecosystem.

Thinking about it, I can't think of any library that uses option 3. Probably in big part to avoid boilerplate and code duplication. If `#[inherent]` (from the linked crate) would be a compiler build in, it would likely be more common.

---

> We make all traits public and provide the inherent impls. Users never need to use the prelude, but we are left with writing boilerplate inherent impls.

The current codebase already implements several traits using `macro_rules!`. (The AST structure of Marker heavily relies on this.) Implementing the functions on the types themselves shouldn't be a big problem with this setup. So I believe, the boilerplate would be manageable, and if not, we could use a proc macro, like the linked one.

---

I've tried to google this question, but wasn't even sure how to phrase it correctly in "generic" terms. :sweat_smile:

For the `Spanned` trait, it feels like having both would be the best solution. But I'll leave the other traits and trait implementations as they are right now. If that also sounds reasonable to you?

Reply:
Yes, sounds reasonable :+1: